Support in app purchases

Benjamin1333 opened the pull request:
- [ ]  release new version

Benjamin1333 committed:
Upgrade outdated dependencies

Benjamin1333 committed:
Support v1 & v2 connect api

Benjamin1333 committed:
Support limit for get requests

Benjamin1333 committed:
Support limit for get requests

Benjamin1333 committed:
Support json map & array data properties

Benjamin1333 committed:
Introduce some iap features

Benjamin1333 committed:
Print error information

kaptnkoala committed:
Merge branch 'main' into feature/iap

# Conflicts:
#	lib/src/api.dart
#	pubspec.lock
#	pubspec.yaml

kaptnkoala commented on lib/src/client.dart:
```diff
@@ -4,7 +4,15 @@ import 'package:appstore_connect/src/model/model.dart';
 import 'package:http/http.dart';
 
-const _apiUri = 'https://api.appstoreconnect.apple.com/v1/';
+extension AppStoreConnectUri on Uri {
+  static Uri v1({String? resource = ''}) {
+    return Uri.parse('https://api.appstoreconnect.apple.com/v1/$resource');
+  }
+
+  static Uri v2({String? resource = ''}) {
+    return Uri.parse('https://api.appstoreconnect.apple.com/v2/$resource');
+  }
+}
```
[Anmerkung] Da diese beide Methoden nur zusätzliche statische Methoden bereitstellen, würde ich sie nicht als einen Extension von Uri modellieren, da sie ja eigentlich gar nicht extenden. Vielleicht eher in einer abstrakten Klassen den wir dann als Namespace missbrauchen.

`resource` würde ich dann nicht als optionalen benannten Parameter sondern als optionalen positionalen Parameter abbilden. Das ist bei vielen Dart Methoden/Klassen so, die nur einen Parameter annehmen und nur eine Funktion haben. Dann spart sich das angeben von Resource.

Benjamin1333 replied:
Done.
Ich hatte die Parameter erst als optionale positional Parameter. Leider muss man dann beim Initialisieren ohne Resource 'null' übergeben. Das fand ich damals nicht so schön. Kannst es dir mal [anschauen](https://github.com/programmierbar/appstore_connect/commit/bb7da3f46f2f33da112f747aa4916771106c1b48).

kaptnkoala replied:
Einen optionalen positionalen Parameter kannst du mit eckigen Klammern angeben, dann braucht du nicht null übergeben, wenn du es leer lassen willst. z.B. `v1([String? resource])`

kaptnkoala commented on lib/src/api.dart:
```diff
@@ -67,4 +74,244 @@ class AppStoreConnectApi {
+      'data': {
+        'type': 'inAppPurchases',
+        'attributes': {
+          'familySharable': false,
+          'inAppPurchaseType': type.toString(),
+          'name': name,
+          'productId': productId,
+          "reviewNote": reviewNote,
```
[Anmerkung] Hier sollten wir überlegen, ob wir in einem öffentlichen Paket nicht alle Attribute vom Anwender konfigurierbar halten wollen. Das `familyShareable` sollten wir z.B. nicht immer auf `false` setzen. Diese könnten wir ähnlich zu den `AppStoreVersionAttributes` halten.

Wenn wir diesen Ansatz verfolgen könnten wahrscheinlich alle der jetzt fix implementierten Requests die `postModel` API des Clients nutzen und die Anpassung für andere Anwendungsfälle erlauben.

Benjamin1333 replied:
done

kaptnkoala commented on lib/src/api.dart:
```diff
@@ -67,4 +74,244 @@ class AppStoreConnectApi {
+    try {
+      _client.putBinary(target, binaryAsset, headers);
+    }catch(e) {
+      _client.putBinary(target, binaryAsset, headers);
```
[Frage] Hier machst du einfach einen weiteren Retry für den Upload?

Benjamin1333 replied:
Ja. Mies, gell. Problem ist: Manchmal sind die UploadOperations nicht 'ready' und brechen dann mit Fehler ab. Ähnliches Problem hatte ich bei hatte ich [zwei Zeilen weiter unten](https://github.com/programmierbar/appstore_connect/blob/bb7da3f46f2f33da112f747aa4916771106c1b48/lib/src/api.dart#L283). Da hab ich es durch ein sleep gelöst. Gefällt mir auch nicht, weiß nur nicht wie man es besser machen kann. Die API suggeriert mir, dass alles ready ist und dann schlägt es manchmal fehl.

kaptnkoala commented on lib/src/model/in_app_purchase.dart:
```diff
@@ -0,0 +1,178 @@
+}
+
+class InAppPurchaseState {
+  static const approved = InAppPurchaseState._('APPROVED');
+  static const developerActionNeeded = InAppPurchaseState._('DEVELOPER_ACTION_NEEDED');
+  static const developerRemovedFromSale = InAppPurchaseState._('DEVELOPER_REMOVED_FROM_SALE');
+  static const inReview = InAppPurchaseState._('IN_REVIEW');
+  static const missingMetadata = InAppPurchaseState._('MISSING_METADATA');
+  static const pendingBinaryApproval = InAppPurchaseState._('PENDING_BINARY_APPROVAL');
+  static const processingContent = InAppPurchaseState._('PROCESSING_CONTENT');
+  static const readyToSubmit = InAppPurchaseState._('READY_TO_SUBMIT');
+  static const rejected = InAppPurchaseState._('REJECTED');
+  static const removedFromSale = InAppPurchaseState._('REMOVED_FROM_SALE');
+  static const waitingForReview = InAppPurchaseState._('WAITING_FOR_REVIEW');
+  static const waitingForUpload = InAppPurchaseState._('WAITING_FOR_UPLOAD');
```
[Hinweis] Mein Code ist hier ein schlechte Referenz gewesen, den InAppPurchaseState hätte man auch gut als Enum abbilden können. Vielleicht hast du es aber auch probiert und es hat aus irgendeinem Grund nicht geklappt.

Benjamin1333 replied:
Hatte ich erst als enum modelliert. Ich glaube die aktuelle mind. Dart Version unterstützt keine Enums. Oder keine Enums mit selbst definierte Value. Bin mir gerade nicht mehr sicher. Da der Rest ebenfalls so umgesetzt war, hab ich nicht groß nachgedacht und es gleich umgesetzt.

Würden wir eher anpassen wenn wir das Paket mal wieder grundsätzlich anfassen?

kaptnkoala replied:
Bin mir nicht sicher, warum ich das damals so gebaut habe. Kann gut sein, dass es Dart damals noch nicht unterstützt hat. Lass uns jetzt einfach die minimale Dart Version hochsetzen. Aber gerne zu einem späteren Zeitpunkt anpassen.

kaptnkoala commented on lib/src/model/in_app_purchase.dart:
```diff
@@ -0,0 +1,178 @@
+
+}
+
+class UploadOperations {
```
[Anmerkung] Ist vom Naming eher ein einzelne UploadOperation.

Benjamin1333 replied:
Stimmt. Done.

kaptnkoala commented on lib/src/model/model.dart:
```diff
@@ -62,10 +75,18 @@ class ModelRelationship {
-    return modelList.cast<T>();
+    final Iterable<Model> modelList;
+    final dataValue = envelope['data'];
+    if (dataValue is Map) {
+      modelList = dataValue.values.map((value) => _parseModel(client, value, includedModels));
+    } else if (dataValue is List) {
```
[Frage] Wann verwenden wir die parseList Methode zum Auspacken einer Map Struktur. Stehen dann in den Keys nur unwichtige Informationen drinne?

Benjamin1333 replied:
Vor diesem PR wurde davon ausgegangen, dass `data` als Map Struktur vorliegt. Die Endpunkte die ich angesprochen habe, liefern allerdings immer ein Array. Deswegen hab ich die Weiche eingebaut. Wenn ich mir den main Branch anschaue verstehe ich nicht wieso der Code funktionieren soll. Beispielsweise das Laden der [AppStoreVersionLocalizations](https://developer.apple.com/documentation/appstoreconnectapi/appstoreversionlocalizationsresponse) liefert ebenfalls ein Array. Keine Ahnung ob ich alles richtig verstehe.

kaptnkoala commented on lib/src/model/in_app_purchase.dart:
```diff
@@ -0,0 +1,178 @@
+        currency = attributes['currency'],
+        super(type, id);
+
+}
```
[Anmerkung] Vielleicht könnten ein paar Models hieraus auch in eigenen Dateien leben.

Benjamin1333 replied:
Ich hab Territory in ein eigenes File verschoben. Das macht Sinn. Die restlichen Models sind Teil des iap "Namespaces" und ich würde sie in dem File erwarten. Oder hab ich was übersehen?

UploadOperation & RequestHeader sind eher Hilfsklassen und erben nicht von Model. Hätte sie dementsprechend nicht in ein anderes File verschoben.

kaptnkoala committed:
Check code style

Benjamin1333 committed:
Rename to singular

Benjamin1333 committed:
Move Territory model into separate file

Benjamin1333 committed:
Remove uri extension & named parameters

Benjamin1333 committed:
Use attributes classes

Benjamin1333 committed:
Use optional positional parameter

Benjamin1333 committed:
Use optional positional parameter

Benjamin1333 committed:
Merge remote-tracking branch 'origin/feature/iap' into feature/iap

Benjamin1333 committed:
Formatting

Benjamin1333 committed:
Fix relation handling

Benjamin1333 committed:
Fix postInAppPurchasePriceSchedule api call

Benjamin1333 committed:
Remove debug log

Benjamin1333 committed:
Add missing model classes

AppStoreConnectClient.postModel will parse the response and requires therefore the registration of result model classes

Benjamin1333 committed:
Fix type

Benjamin1333 committed:
Add todo

kaptnkoala commented on lib/src/model/model.dart:
```diff
@@ -62,17 +66,53 @@ abstract class ModelAttributes {
   }
 }
 
+class MultipleModelRelationship extends ModelRelationship {
```
[Anmerkung] Lass uns das MultiModelRelationship nennen. Multiple ist so sperrig.

Benjamin1333 replied:
done

kaptnkoala commented on lib/src/api.dart:
```diff
@@ -80,228 +79,134 @@ class AppStoreConnectApi {
-
-    await _client.post(AppStoreConnectUri.v2(resource: 'inAppPurchases'), body);
+  Future<bool> postInAppPurchase(InAppPurchaseAttributes attributes) async {
+    await _client.postModel(
+      AppStoreConnectUri.v2(),
+      'inAppPurchases',
+      attributes: attributes,
+      relationships: {
+        'app': SingleModelRelationship(type: 'apps', id: _appId),
+      },
+    );
```
[Anmerkung] Der Post würde dir ja in der Response schon das erstellte InAppPruchase Objekt zurückgeben. Wenn du das hier zurückgibst, musst du im Mate Command nicht wieder alle InAppPurchases laden um anhand der Product ID die IAP ID aufzulösen.

Benjamin1333 replied:
Gute Idee. done.

kaptnkoala commented on lib/src/api.dart:
```diff
@@ -80,228 +79,134 @@ class AppStoreConnectApi {
-    };
-
-    await _client.post(AppStoreConnectUri.v1('inAppPurchaseLocalizations'), body);
+  Future<bool> postInAppPurchaseLocalization(
```
[Anmerkung] Diese und auch alle weitere Methoden würde ich an InAppPurchase hinterlegen, damit man mit einer mehr objektorientierten API arbeiten kann.

Benjamin1333 replied:
done

Benjamin1333 committed:
Renaming

Benjamin1333 committed:
Return created InAppPurchase

Benjamin1333 committed:
Remove strange static field

Benjamin1333 committed:
Refactor iap classes to callable models

Benjamin1333 committed:
Remove deprecated field

kaptnkoala reviewed (approved):
Wunderschön! Ich liebe es!! ❤️ ❤️ ❤️

Benjamin1333 commented:
Das freut mich sehr.
